[Storage] azure-storage-file-datalake

XiaoningLiu opened the pull request:
Initial code for Azure Storage DataLake package. It's kind of big. Recommend to review using [CodeFlow extension](https://chrome.google.com/webstore/detail/codeflow/aphnoipocoffpdafmiidfmaiadhilelm)

I will send out a code flow CR and keep align with this PR too for convenience.

XiaoningLiu commented:
@jiacfan @jeremymeng @HarshaNalluru @bterlson Please take a look. Currently this PR includes main feature implementation. And will update with following implementations into the PR.

- [X] Feature
- [ ] Test cases
- [ ] Mock test
- [ ] Bundle scripts
- [ ] Documents (Readme.md, ChangeLog.md, BreakingChange.md)
- [ ] CI & Pipeline
- [ ] Samples (May move into next release)
- [ ] Comments
- [ ] Others

jiacfan commented on sdk/storage/storage-file-datalake/package.json:
```diff
@@ -0,0 +1,160 @@
+{
+  "name": "@azure/storage-file-datalake",
+  "version": "12.0.0-preview.7",
```
Is this by design?

XiaoningLiu replied:
Should be preview.6 as the part of track2 preview series.

jiacfan commented on sdk/storage/storage-file-datalake/src/AccountSASPermissions.ts:
```diff
@@ -0,0 +1,169 @@
+  public write: boolean = false;
+
+  /**
+   * Permission to create blobs and files granted.
```
delete

jiacfan commented on sdk/storage/storage-file-datalake/src/AccountSASPermissions.ts:
```diff
@@ -0,0 +1,169 @@
+
+  /**
+   * Permission to create blobs and files granted.
+   *
```
General which applies to other fields, what's the document to support adls gen 2 account SAS. Existing things are related to blob. And this SDK is possibly for data lake user only?

XiaoningLiu replied:
There is no document. Basically, dfs endpoint applies all blob sas.

jiacfan commented on sdk/storage/storage-file-datalake/src/DataLakeDirectoryClient.ts:
```diff
@@ -0,0 +1,76 @@
+  ): Promise<PathCreateResponse>;
+  public async create(options?: DirectoryCreateOptions): Promise<DirectoryCreateResponse>;
+  public async create(
+    resourceTypeOrOptions?: PathResourceType | PathCreateOptions,
```
DirectoryCreateOptions?

XiaoningLiu replied:
DirectoryCreateOptions is an alias of PathCreateOptions

XiaoningLiu replied:
Will update

jiacfan commented on sdk/storage/storage-file-datalake/src/DataLakeFileClient.ts:
```diff
@@ -0,0 +1,191 @@
+  ): Promise<PathCreateResponse>;
+  public async create(options?: FileCreateOptions): Promise<FileCreateResponse>;
+  public async create(
+    resourceTypeOrOptions?: PathResourceType | PathCreateOptions,
```
FileCreateOptions?

XiaoningLiu replied:
FileCreateOptionsis an alias of PathCreateOptions. Will update anyway.

XiaoningLiu commented:
All checks are green : )

jeremymeng commented on sdk/storage/storage-file-datalake/README.md:
```diff
@@ -0,0 +1,520 @@
+- A _file system_ in the storage account used via `DataLakeFileSystemClient`
+- A _path_ in a file system used via `DataLakeDirectoryClient` or `DataLakeFileClient`
+
+|ADLS Gen2 	                | Blob       |
```
>ADLS [](start = 1, length = 4)

nit: ADLS might not be an obvious acronym. We should introduce it first (like HNS below)

XiaoningLiu replied:
Will do.

jeremymeng commented on sdk/storage/storage-file-datalake/README.md:
````diff
@@ -0,0 +1,520 @@
+main();
+```
+
+### List the containers
````
>containers [](start = 13, length = 10)

Should `containers` be file systems instead?

XiaoningLiu replied:
Yes, updated.

jeremymeng commented on sdk/storage/storage-file-datalake/README.md:
```diff
@@ -0,0 +1,520 @@
+
+### List the containers
+
+Use `DataLakeServiceClient.listFileSystems()` function to iterate the containers,
```
>containers [](start = 70, length = 10)

same as above, containers should be file systems?

XiaoningLiu replied:
Yes, should be file system. Technically, they are the same. But in datalake SDK, we call it as file system. Updated.

jeremymeng commented:
>   "dependencies": {

This file probably needs more work later. Can we at least use MIT licence?

---
Refers to: sdk/storage/storage-file-datalake/samples/package.json:12 in 49e1edd. [](commit_id = 49e1eddea777be6341077066bf70922a8c86a4fe, deletion_comment = False)

jeremymeng commented:
>   // List containers

I know it's from other samples, but please delete this line

---
Refers to: sdk/storage/storage-file-datalake/samples/javascript/basic.js:35 in 49e1edd. [](commit_id = 49e1eddea777be6341077066bf70922a8c86a4fe, deletion_comment = False)

jeremymeng commented:
>   const serviceClient = new DataLakeServiceClient(

delete // List containers line

---
Refers to: sdk/storage/storage-file-datalake/samples/typescript/basic.ts:36 in 49e1edd. [](commit_id = 49e1eddea777be6341077066bf70922a8c86a4fe, deletion_comment = False)

XiaoningLiu commented:
> > // List containers
> 
> I know it's from other samples, but please delete this line
> 
> Refers to: sdk/storage/storage-file-datalake/samples/javascript/basic.js:35 in [49e1edd](https://github.com/Azure/azure-sdk-for-js/commit/49e1eddea777be6341077066bf70922a8c86a4fe). [](commit_id = [49e1edd](https://github.com/Azure/azure-sdk-for-js/commit/49e1eddea777be6341077066bf70922a8c86a4fe), deletion_comment = False)

Updated. FYI, I added basic sample in this preview release. Can add more samples in following releases.

jiacfan commented on sdk/storage/storage-file-datalake/BreakingChanges.md:
```diff
@@ -0,0 +1,5 @@
+
+## 2019.12 12.0.0-preview.6
+
+- Initial Release. API version 2019-02-02 supported. Please see the README for information on the new design.
```
Please ensure if 2019-02-02 is wished.

XiaoningLiu replied:
Yes.

jiacfan commented on sdk/storage/storage-file-datalake/README.md:
```diff
@@ -0,0 +1,520 @@
+Key Features of DataLake Storage Gen2 include:
+
+- Hadoop compatible access
+- A super set of POSIX permissions
```
minor: some ends with '.' some without, would better keep align

XiaoningLiu replied:
Thanks! For short words we don't have dot ending. For long sentences will end with dot.

jiacfan commented on sdk/storage/storage-file-datalake/README.md:
```diff
@@ -0,0 +1,520 @@
+- A _path_ in a file system used via `DataLakeDirectoryClient` or `DataLakeFileClient`
+
+|Azure DataLake Gen2 	      | Blob       |
+| --------------------------| ---------- |
```
Please ensure alignment

XiaoningLiu replied:
Updated. Alignment here doesn't influence markdown rendering

jiacfan commented on sdk/storage/storage-file-datalake/README.md:
```diff
@@ -0,0 +1,520 @@
+
+- **Recommended way to instantiate a `DataLakeServiceClient` - with `DefaultAzureCredential` from `@azure/identity` package**
+
+  Setup : Reference - Authorize access to blobs and queues with Azure Active Directory from a client application - https://docs.microsoft.com/azure/storage/common/storage-auth-aad-app
```
adls gen2 should has it's own page for ACL and OAuth usage.

XiaoningLiu replied:
Actually, these docs are appliable for adls, it reuse blob related roles, see

![image](https://user-images.githubusercontent.com/6889160/70196878-01e21f00-1745-11ea-8382-7d87dcd9d1a8.png)

XiaoningLiu replied:
Added a notice about datalake reuses blob permissions during OAuth.

jiacfan commented on sdk/storage/storage-file-datalake/README.md:
```diff
@@ -0,0 +1,520 @@
+  await fileClient.create();
+  await fileClient.append(content, 0, content.length);
+  await fileClient.flush(content.length);
+  console.log(`Create and upload file${fileName} successfully`);
```
minor: space missed after file?

XiaoningLiu replied:
Updated.

jiacfan commented:
>   "description": "Copy the sample file you want to a folder of your choice and run the following command in your samples folder to install the npm package for Azure Storage DataLake library.",

test command?

---
Refers to: sdk/storage/storage-file-datalake/samples/package.json:4 in 0da104a. [](commit_id = 0da104ab9e6575568321e732990747322f743523, deletion_comment = False)

jiacfan commented:
>   console.log("deleted filesystem");

minor: Deleted

---
Refers to: sdk/storage/storage-file-datalake/samples/javascript/basic.js:81 in 0da104a. [](commit_id = 0da104ab9e6575568321e732990747322f743523, deletion_comment = False)

jiacfan commented:
>   console.log("deleted filesystem");

minor: Deleted

---
Refers to: sdk/storage/storage-file-datalake/samples/typescript/basic.ts:81 in 0da104a. [](commit_id = 0da104ab9e6575568321e732990747322f743523, deletion_comment = False)

jiacfan commented:
>  Setup: Enter your storage account name and shared key in main()

Inaccurate, it's reading from env var.

---
Refers to: sdk/storage/storage-file-datalake/samples/typescript/basic.ts:2 in 0da104a. [](commit_id = 0da104ab9e6575568321e732990747322f743523, deletion_comment = False)

XiaoningLiu commented:
> > Setup: Enter your storage account name and shared key in main()
> 
> Inaccurate, it's reading from env var.
> 
> Refers to: sdk/storage/storage-file-datalake/samples/typescript/basic.ts:2 in [0da104a](https://github.com/Azure/azure-sdk-for-js/commit/0da104ab9e6575568321e732990747322f743523). [](commit_id = [0da104a](https://github.com/Azure/azure-sdk-for-js/commit/0da104ab9e6575568321e732990747322f743523), deletion_comment = False)

It can read from environment variables OR from inline constant string. Environment variable is usually used by us for testing purpose.

XiaoningLiu commented:
> > console.log("deleted filesystem");
> 
> minor: Deleted
> 
> Refers to: sdk/storage/storage-file-datalake/samples/typescript/basic.ts:81 in [0da104a](https://github.com/Azure/azure-sdk-for-js/commit/0da104ab9e6575568321e732990747322f743523). [](commit_id = [0da104a](https://github.com/Azure/azure-sdk-for-js/commit/0da104ab9e6575568321e732990747322f743523), deletion_comment = False)

Updated
